### PR TITLE
feat: make `glm.disconnect` `rentalPool.drainAndClear` and `rental.stopAndFinalize` idempotent

### PR DESCRIPTION
**Pull request by SewerynKras**
- make `glm.disconnect` idempotent
- make `pool.drainAndClear` idempotent
- make `rental.stopAndFinalize` idempotent
- interrupt `oneOf` and `manyOf` if `glm` gets disconnected while it's being created
- interrupt rentals being created in the pool when it gets drained

### CODE REVIEW

File: src/golem-network/golem-network.test.ts [hunk 1]
```diff
@@ -187,4 +187,19 @@ describe("Golem Network", () => {
...
+    it("reuses the same promise if called multiple times", async () => {
+      const glm = getGolemNetwork();
+      const glmSpy = spy(glm);
+      when(glmSpy["startDisconnect"]()).thenResolve();
+      expect(glm["disconnectPromise"]).toBeUndefined();
+      const promise1 = glm.disconnect();
+      const promise2 = glm.disconnect();
+      const promise3 = glm.disconnect();
+      expect(glm["disconnectPromise"]).toBeDefined();
+      await Promise.all([promise1, promise2, promise3]);
```
Comment (mgordel) [inline src/golem-network/golem-network.test.ts:200]: Parallel runs passes the test, but sequential fails:
```ts
await promise1;
await promise2;
```

File: src/golem-network/golem-network.ts [hunk 2]
```diff
@@ -314,20 +315,35 @@ export class GolemNetwork {
...
-
-    this.services.proposalCache.flushAll();
-
-    this.events.emit("disconnected");
-    this.hasConnection = false;
+    if (this.disconnectPromise) {
+      return this.disconnectPromise;
+    }
+    this.disconnectPromise = this.startDisconnect().finally(() => {
+      this.disconnectPromise = undefined;
```
Comment (mgordel) [inline src/golem-network/golem-network.ts:344]: why is this needed? this means that running disconnect sequentially will not be idempotent
```ts
await glm.disconnect();
await glm.disconnect(); // runs disconnect again causes another attempt to releas allocation etc.
```
so maybe it should be `undefined` only when the user run the `connect` again?

File: src/resource-rental/resource-rental-pool.ts [hunk 3]
```diff
@@ -266,20 +289,14 @@ export class ResourceRentalPool {
...
-    this.highPriority.clear();
-    this.borrowed.clear();
-    this.isDraining = false;
-    this.events.emit("end");
-    return;
+    if (this.isDraining) {
+      return this.drainPromise;
+    }
+    this.drainPromise = this.startDrain().finally(() => {
+      this.drainPromise = undefined;
```
Comment (mgordel) [inline src/resource-rental/resource-rental-pool.ts:296]: same behavior as above

File: src/golem-network/golem-network.test.ts (hunk 1 above)
Comment (SewerynKras) [reply src/golem-network/golem-network.test.ts:200]: This is the correct logic in my opinion, if you await the promise then the disconnect finishes and on the next run it's the same as running `disconnect` on a newly created `GolemNetwork` (without `connect` first), nothing should happen in that case

File: src/golem-network/golem-network.ts (hunk 2 above)
Comment (SewerynKras) [reply src/golem-network/golem-network.ts:344]: `cleanupTasks` should be set to `null` after disconnect finishes, this is a bug

File: src/golem-network/golem-network.test.ts (hunk 1 above)
Comment (SewerynKras) [reply src/golem-network/golem-network.test.ts:200]: What I mean is I think this should be valid:
```ts
const glm = new GolemNetwork()
await glm.connect();
await glm.disconnect();
await glm.disconnect() // nothing happens
await glm.connect();
await glm.disconnect();
```

Comment (mgordel) [reply src/golem-network/golem-network.test.ts:200]: ok, I agree, so disconnecting on an unconnected `GolemNetwork` should not result in any action on the network, so it should either be resolved immediately or give an error, e.g. `GolemUserError("Network is not connected")`

Comment (SewerynKras) [reply src/golem-network/golem-network.test.ts:200]: same for rental pool
```ts
const pool = new ResourceRentalPool({ ... });
await pool.ready();
await pool.drainAndClear(); // finalizes all rentals
await pool.drainAndClear(); // nothing happens
await pool.ready();
// pool is ready to use again
await pool.drainAndClear(); // finalizes all rentals
```

**Review by mgordel — APPROVED** [review]
